Composer/packages/client/src/components/CreationFlow/LocationBrowser/FileSelector.tsx

```diff
@@ -46,52 +62,147 @@ type SortState = {
-      <Link
-        aria-label={
-          file.name === '..'
-            ? formatMessage('previous folder')
-            : formatMessage('{icon} name is {file}', {
-                icon: iconName,
-                file: file.name,
-              })
-        }
-        onClick={() => onFileChosen(file)}
-      >
-        {file.name}
-      </Link>
-    </div>
-  );
-};
-
 export const FileSelector: React.FC<FileSelectorProps> = (props) => {
   const {
     onFileChosen,
     focusedStorageFolder,
     checkShowItem,
     onCurrentPathUpdate,
     operationMode,
+    createFolder,
+    updateFolder,
     isWindows = false,
+    storages,
   } = props;
   // for detail file list in open panel
-  const currentPath = path.join(focusedStorageFolder.parent, focusedStorageFolder.name);
+  const [currentPath, setCurrentPath] = useState(path.join(focusedStorageFolder.parent, focusedStorageFolder.name));
+  const initialPath = useRef(path.join(focusedStorageFolder.parent, focusedStorageFolder.name)).current;
+  const currentStorageIndex = useRef(0);
+  const storage = storages[currentStorageIndex.current];
+  const storageId = storage.id;
+  const [selectedIndex, setSelectedIndex] = useState(-1);
+  const [folderName, setFolderName] = useState('');
+  const [editMode, setEditMode] = useState(EditMode.NONE);
+  const [nameError, setNameError] = useState('');
+
+  const createOrUpdateFolder = async (index: number) => {
+    const isValid = nameRegex.test(folderName);
+    const isDup = storageFiles.some((file) => file.name === folderName) && storageFiles[index].name !== folderName;
+    if (isValid && !isDup) {
+      if (editMode === EditMode.Creating) {
+        createFolder && (await createFolder(initialPath, folderName));
+        await onCurrentPathUpdate(path.join(initialPath, folderName), storageId);
+      }
+      if (editMode === EditMode.Updating) {
+        updateFolder && (await updateFolder(initialPath, storageFiles[index].name, folderName));
+        await onCurrentPathUpdate(initialPath, storageId);
+      }
+      setEditMode(EditMode.NONE);
+      setNameError('');
+      setFolderName('');
+    } else if (!folderName) {
+      // an empty name means to cancel the operation
+      cancelEditOperation();
+    } else if (isDup) {
+      const nameError = formatMessage('folder {folderName} already exists', { folderName });
+      setNameError(nameError);
+      setFolderName('');
+    } else {
+      const nameError = formatMessage('Spaces and special characters are not allowed.');
+      setNameError(nameError);
+      setFolderName('');
+    }
+  };
+
+  const onEditButtonClick = (file: File) => {
+    setEditMode(EditMode.Updating);
+    setFolderName(file.name);
+  };
+
+  const cancelEditOperation = () => {
+    setEditMode(EditMode.NONE);
+    setNameError('');
+    setCurrentPath(initialPath);
+    setFolderName('');
+  };
+
+  const handleKeydown = (e, index) => {
+    if (e.key === 'Enter' && index) {
```

Comment:
What if `index` is 0?

Reply:
This is on purpose. The folder[0] is the go to parent folder button. The handleKeydown function should never be triggered on index 0.

Reply:
Let's make that more explicit in the code. `index > 0` with a comment explaining that.

Reply:
done